src/Dfc.CourseDirectory.WebV2/Features/Venues/AddVenue/Address.cs

```diff
@@ -0,0 +1,116 @@
+        public string Postcode { get; set; }
+    }
+
+    public class Handler :
+        IRequestHandler<Query, Command>,
+        IRequestHandler<Command, OneOf<ModelWithErrors<Command>, Success>>
+    {
+        private readonly JourneyInstance<AddVenueJourneyModel> _journeyInstance;
+        private readonly ISearchClient<Onspd> _onspdSearchClient;
+
+        public Handler(JourneyInstanceProvider journeyInstanceProvider, ISearchClient<Onspd> onspdSearchClient)
+        {
+            _journeyInstance = journeyInstanceProvider.GetInstance<AddVenueJourneyModel>();
+            _onspdSearchClient = onspdSearchClient;
+        }
+
+        public Task<Command> Handle(Query request, CancellationToken cancellationToken)
+        {
+            ThrowIfFlowStateNotValid();
+
+            return Task.FromResult(new Command()
+            {
+                AddressLine1 = _journeyInstance.State.AddressLine1,
+                AddressLine2 = _journeyInstance.State.AddressLine2,
+                Town = _journeyInstance.State.Town,
+                County = _journeyInstance.State.County,
+                Postcode = _journeyInstance.State.Postcode
+            });
+        }
+
+        public async Task<OneOf<ModelWithErrors<Command>, Success>> Handle(Command request, CancellationToken cancellationToken)
+        {
+            ThrowIfFlowStateNotValid();
+
+            var validator = new CommandValidator();
+            var validationResult = await validator.ValidateAsync(request);
+
+            if (!validationResult.IsValid)
+            {
+                return new ModelWithErrors<Command>(request, validationResult);
+            }
+
+            var onspdSearchResult = await _onspdSearchClient.Search(new OnspdSearchQuery() { Postcode = request.Postcode });
+
+            if (onspdSearchResult.Items.Count == 0)
+            {
+                validationResult = new ValidationResult(new[]
+                {
+                    new ValidationFailure(nameof(request.Postcode), "Enter a valid postcode")
+                });
+
+                return new ModelWithErrors<Command>(request, validationResult);
+            }
+
+            var onspdPostcodeRecord = onspdSearchResult.Items.Single();
+
+            _journeyInstance.UpdateState(state =>
+            {
+                state.AddressLine1 = request.AddressLine1;
+                state.AddressLine2 = request.AddressLine2;
+                state.Town = request.Town;
+                state.County = request.County;
+                state.Postcode = request.Postcode;
+                state.Latitude = onspdPostcodeRecord.Record.lat;
+                state.Longitude = onspdPostcodeRecord.Record.@long;
+                state.AddressIsOutsideOfEngland = onspdPostcodeRecord.Record.Country != "England";
+                state.ValidStages |= AddVenueCompletedStages.Address;
+            });
+
+            return new Success();
+        }
+
+        private void ThrowIfFlowStateNotValid()
```

Comment:
Unnecessary abstraction

Reply:
It's to be consistent with other pages.

Reply:
I don't think consistency with a different class is an acceptable argument - why does it need to be consistent? In this particular case all it does it make the code more difficult to read.